src/sql/src/query_model/rewrite_engine.rs

```diff
@@ -999,3 +1000,168 @@ impl Rule for GroupingToDistinct {
         b.distinct = DistinctOperation::Enforce;
     }
 }
+
+/// Any Scalar quantifier whose input box that either:
+/// 1) produces at most one row AND has a predicate that discards null rows.
+/// 2) produces exactly one row.
+/// can be converted into a Foreach quantifier
+#[derive(Default)]
+struct ScalarToForeach {
```

Comment:
As a rule, once `action` returns nothing guarantees the cached information in the rule to be valid anymore, since another transform may make that cached information invalid by altering the graph.

The way I'd implement this is by adding two recursive methods in `QueryBox`

```
let b = model.get_box(box_id);
self.to_convert = b.quantifiers.iter().filter_map(|q_id| {
    let q = model.get_quantifier(*q_id);
    if q.quantifier_type = QuantifierType::Scalar {
       Some((*q_id, q.input_box))
    } else {
        None
    }
})
.filter_map(|(q_id, input_box)| {
    let i_b = model.get_box(input_box);
    if i_b.exactly_one_row() {
        return Some(*q_id);
    } else if i_b.one_row_at_most() {
         // find out whether the outer box has a null rejecting predicate on a column from q_id quantifier
    }
    None
}).collect();
```

Reply:
Ah, in that case I think that the rule object should be destroyed after `action` returns, and a fresh rule object should be created each time we call `apply_rule`. I have added a commit where I make a prototype RuleGenerator that generates a blank rule object each time we call `apply_rule`.

Reply:
oh, that's really nice!